tokio/src/io/util/read_to_end.rs

```diff
@@ -12,81 +12,136 @@ use std::task::{Context, Poll};
-        unsafe {
-            self.buf.set_len(self.len);
+/// SAFETY: Before first calling this method, the unused capacity must have been
+/// prepared for use with the provided AsyncRead. This can be done using the
+/// `prepare_buffer` function later in this file.
+pub(super) unsafe fn read_to_end_internal<R: AsyncRead + ?Sized>(
+    buf: &mut Vec<u8>,
+    mut reader: Pin<&mut R>,
+    num_read: &mut usize,
+    cx: &mut Context<'_>,
+) -> Poll<io::Result<usize>> {
+    loop {
+        // safety: The caller promised to prepare the buffer.
+        let ret = ready!(poll_read_to_end(buf, reader.as_mut(), cx));
+        match ret {
+            Err(err) => return Poll::Ready(Err(err)),
+            Ok(0) => return Poll::Ready(Ok(mem::replace(num_read, 0))),
+            Ok(num) => {
+                *num_read += num;
+            }
         }
     }
 }
 
-// This uses an adaptive system to extend the vector when it fills. We want to
-// avoid paying to allocate and zero a huge chunk of memory if the reader only
-// has 4 bytes while still making large reads if the reader does have a ton
-// of data to return. Simply tacking on an extra DEFAULT_BUF_SIZE space every
-// time is 4,500 times (!) slower than this if the reader has a very small
-// amount of data to return.
-//
-// Because we're extending the buffer with uninitialized data for trusted
-// readers, we need to make sure to truncate that if any of this panics.
-pub(super) fn read_to_end_internal<R: AsyncRead + ?Sized>(
-    mut rd: Pin<&mut R>,
-    cx: &mut Context<'_>,
+/// Tries to read from the provided AsyncRead.
+///
+/// The length of the buffer is increased by the number of bytes read.
+///
+/// SAFETY: Before first calling this method, the unused capacity must have been
+/// prepared for use with the provided AsyncRead. This can be done using the
+/// `prepare_buffer` function later in this file.
+unsafe fn poll_read_to_end<R: AsyncRead + ?Sized>(
     buf: &mut Vec<u8>,
-    start_len: usize,
+    read: Pin<&mut R>,
+    cx: &mut Context<'_>,
 ) -> Poll<io::Result<usize>> {
-    let mut g = Guard {
-        len: buf.len(),
-        buf,
+    // This uses an adaptive system to extend the vector when it fills. We want to
+    // avoid paying to allocate and zero a huge chunk of memory if the reader only
+    // has 4 bytes while still making large reads if the reader does have a ton
+    // of data to return. Simply tacking on an extra DEFAULT_BUF_SIZE space every
+    // time is 4,500 times (!) slower than this if the reader has a very small
+    // amount of data to return.
+    reserve(buf, &*read, 32);
+
+    let unused_capacity: &mut [MaybeUninit<u8>] = get_unused_capacity(buf);
+
+    // safety: The buffer has been prepared for use with the AsyncRead before
+    // calling this function.
+    let slice: &mut [u8] = {
+        let ptr = unused_capacity.as_mut_ptr().cast::<u8>();
+        let len = unused_capacity.len();
+        std::slice::from_raw_parts_mut(ptr, len)
```

Comment:
For my own understanding: this looks equivalent to `transmute`-ing the uninitialized slice to an initialized slice, but with more steps. My guess is that we're doing this instead because it is less powerful than `transmute`, and can't (for instance) reinterpret the length field as some entirely distinct type accidentally?

Reply:
Yes, it's just a transmute. I'm doing it like this because using transmutes for pointer casts is usually rather strongly discouraged. At a bit more thought, we can also do it like this:
```rust
let slice: &mut [u8] = {
    &mut *(unused_capacity as *mut [MaybeUninit<u8>] as *mut [u8])
};
```